// babel-plugin-transform-adana/dist/tags.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.applyRules = applyRules;
exports.extract = extract;
exports.addRules = addRules;

var _meta = require('./meta');

var _meta2 = _interopRequireDefault(_meta);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var COMMENT_PATTERN = /^\s*@?(adana|coverage|test|istanbul):?\s*(.*)\s*/;

function within(a, b) {
  return a.start.line >= b.start.line && a.start.column >= b.start.column && a.end.line <= b.end.line && a.end.column <= b.end.column;
}

function applyRules(state) {
  var coverage = (0, _meta2.default)(state);
  coverage.entries.forEach(function (entry) {
    var result = {};
    var output = [];
    entry.tags.forEach(function (tag) {
      result[tag] = true;
    });
    coverage.rules.forEach(function (rule) {
      if (within(rule.loc, entry.loc)) {
        result[rule.tag] = rule.value;
      }
    });
    Object.keys(result).forEach(function (tag) {
      var value = result[tag];
      if (value) {
        output.push(tag);
      }
    });
    entry.tags = output;
  });
}

function extract(comment) {
  var output = {};
  var result = COMMENT_PATTERN.exec(comment);
  if (result) {
    var entries = result[2].split(/\s+/);
    entries.forEach(function (entry) {
      switch (entry.charAt(0)) {
        case '+':
          output[entry.substr(1)] = true;
          break;
        case '-':
          output[entry.substr(1)] = false;
          break;
        default:
          break;
      }
    });
  }
  return output;
}

function addRules(state, loc, comments) {
  if (comments) {
    var coverage = (0, _meta2.default)(state);
    comments.forEach(function (comment) {
      var values = extract(comment.value);
      Object.keys(values).forEach(function (tag) {
        coverage.rules.push({
          tag: tag,
          value: values[tag],
          loc: loc
        });
      });
    });
  }
}
